Program/src/port.c

```diff
@@ -0,0 +1,303 @@
+/*------------------------------------------------------------------------------
+This file is protected under copyright. If you want to use it,
+please include this text, that is my only stipulation.  
+
+Author: Evan Giese
+------------------------------------------------------------------------------*/
+#include "port.h"
+
+
+#ifdef POSIX_PORT
+    #include <pthread.h>
+    #include <errno.h>
+    #include <limits.h>
+    #include <stdlib.h>
+    #include <stdio.h>
+    #include <arpa/inet.h>
+    #include <stdarg.h>
+#endif
+
+#ifdef POSIX_FILESYSTEM
+    #include <unistd.h>
+    #include <fcntl.h>
+#endif
+
+#ifdef FREE_RTOS_PORT 
+    #include "FreeRTOS.h"
+    #include "task.h"
+    #include "portable.h"
+#endif
+
+#ifdef CSP_NETWORK
+    #include "csp.h"
+#endif
+
+
+/*------------------------------------------------------------------------------
+    File system port functions, these are used to interchange different 
+    File systems
+------------------------------------------------------------------------------*/
+int ssp_rename(const char *old, const char *new) {
+    #ifdef POSIX_FILESYSTEM
+        return rename(old, new);
+    #endif
+}
+
+int ssp_write(int fd, const void *buf, size_t count) {
+    #ifdef POSIX_FILESYSTEM
+        return write(fd, buf, count);
+    #endif
+}
+
+
+int ssp_read(int fd, char* buff, size_t size) {
+    #ifdef POSIX_FILESYSTEM
+        return read(fd, buff, size);
+    #endif
+
+}
+
+//SEEK_END 2  SEEK_CUR 1  SEEK_SET 0 
+int ssp_lseek(int fd, int offset, int whence) {
+    #ifdef POSIX_FILESYSTEM
+        return lseek(fd, offset, whence);
+    #endif
+} 
+
+int ssp_open(char *pathname, int flags) {
+    #ifdef POSIX_FILESYSTEM
+        //open with read and write permissions
+        return open(pathname, flags, 0666);
+    #endif
+}
+
+int ssp_close(int fd) {
+    #ifdef POSIX_FILESYSTEM
+        return close(fd);
+    #endif
+}
+
```

Comment:
What I could do here Is return the error code by default!

Reply:
Sure, what ever makes sense. If could be that these functions can't be invoked because every they appear there is a check for `POSIX_FILESYSTEM`. However, in future, someone may not realize this is the convention and call one by mistake, then be mystified when nothing happens. So, I'd say it's better programming to at least log something.

Reply:
the POSIX_FILESYSTEM precompiler flag needs to be set upon compilation. We are also using RED_FS (reliance edge) which i haven't included yet. If there is no filesystem, this program will not work (its an FTP app after all). Maybe letting the user not compile unless one flag is set is the better programming method here?

Reply:
Depending on the build system in use, you can put in all kinds of useful compile-time checks (e.g., in CMake checking that packages exist and are > some min version). So, if you can improve the reliability and results by doing stuff like that, I say do it. I really can't tell you exactly what to do as there is no way I know enough; you are the expert :-)